api/configuration/v1alpha1/kongkey_types.go

```diff
@@ -0,0 +1,182 @@
+)
+
+// KongKey is the schema for KongKey API which defines a KongKey entity.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Programmed",description="The Resource is Programmed on Konnect",type=string,JSONPath=`.status.conditions[?(@.type=='Programmed')].status`
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.spec.controlPlaneRef) || has(self.spec.controlPlaneRef)", message="controlPlaneRef is required once set"
+// +kubebuilder:validation:XValidation:rule="(!self.status.conditions.exists(c, c.type == 'Programmed' && c.status == 'True')) ? true : oldSelf.spec.controlPlaneRef == self.spec.controlPlaneRef", message="spec.controlPlaneRef is immutable when an entity is already Programmed"
+type KongKey struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec KongKeySpec `json:"spec"`
+
+	// +kubebuilder:default={conditions: {{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status KongKeyStatus `json:"status,omitempty"`
+}
+
+// GetKonnectStatus returns the Konnect status contained in the KongKey status.
+func (t *KongKey) GetKonnectStatus() *konnectv1alpha1.KonnectEntityStatus {
+	if t.Status.Konnect == nil {
+		return nil
+	}
+	return &t.Status.Konnect.KonnectEntityStatus
+}
+
+// GetKonnectID returns the Konnect ID in the KongKey status.
+func (t *KongKey) GetKonnectID() string {
+	if t.Status.Konnect == nil {
+		return ""
+	}
+	return t.Status.Konnect.ID
+}
+
+// SetKonnectID sets the Konnect ID in the KongKey status.
+func (t *KongKey) SetKonnectID(id string) {
+	if t.Status.Konnect == nil {
+		t.initKonnectStatus()
+	}
+	t.Status.Konnect.ID = id
+}
+
+// GetControlPlaneID returns the ControlPlane ID in the KongKey status.
+func (t *KongKey) GetControlPlaneID() string {
+	if t.Status.Konnect == nil {
+		return ""
+	}
+	return t.Status.Konnect.ControlPlaneID
+}
+
+// SetControlPlaneID sets the ControlPlane ID in the KongKey status.
+func (t *KongKey) SetControlPlaneID(id string) {
+	if t.Status.Konnect == nil {
+		t.initKonnectStatus()
+	}
+	t.Status.Konnect.ControlPlaneID = id
+}
+
+// GetTypeName returns the KongKey Kind name.
+func (t KongKey) GetTypeName() string {
+	return "KongKey"
+}
+
+// GetConditions returns the Status Conditions.
+func (t *KongKey) GetConditions() []metav1.Condition {
+	return t.Status.Conditions
+}
+
+// SetConditions sets the Status Conditions.
+func (t *KongKey) SetConditions(conditions []metav1.Condition) {
+	t.Status.Conditions = conditions
+}
+
+func (t *KongKey) initKonnectStatus() {
+	t.Status.Konnect = &konnectv1alpha1.KonnectEntityStatusWithControlPlaneAndKeySetRef{}
+}
+
+// KongKeySpec defines the spec for a KongKey.
```

Comment:
I believe we should.

OTOH though: there is a corner case here where users could:

- specify first a CP ref and then remove that ref for a keyset ref which belongs to a different CP ref
- specify first the keyset ref and then remove that for a CP ref which is a different one that the one that keyset was bound to

To move ourselves forward I suggest that we only allow setting the CP ref for now and create an issue for implementing keyset ref binding.

I am curious to your thoughts about this.

Reply:
I believe we should require the CP ref to be always set for the `Key` (as that can exist without a `KeySet`) - once set it shouldn't change as it is defined.

If the `KeySet` ref is set, we should ensure in the reconciler it refers to a `KeySet` that is bound to the same `CP` as the `Key`. If it's not, we populate a condition telling the user that the `KeySet` reference is invalid.

Reply:
This approach would allow users to create objects with conflicting CP refs (one direct and the other one indirect form the keyset): this would be the first object that allows this.

I'd suggest we not implement the keyset membership just now and postpone solving this until later. I'm open to other solutions though if they'd prevent admission of resources with conflicting cp refs.

Reply:
@czeslavo This seems like a reasonable approach. Let's move forward with it.

Even though I do not like allowing users to create invalid objects it seems that there's no easy way to prevent this.